### PR TITLE
Avoid assumptions about implementation in EC J-PAKE tests

### LINKED ISSUE #6502: Avoid assumptions about implementation in EC J-PAKE tests
Currently the tests for EC J-PAKE assume that it's OK to tests multiple invalid parameters in a row, and that in some circumstance you can successfully do a valid operation after an invalid one; however that is not necessarily the case, the first test could have put the operation in a special state that affects the result of subsequent tests.

Also, after https://github.com/Mbed-TLS/mbedtls/pull/6390 the tests also rely on the assumption that making a shallow copy of the operation context works, which happens to be true of our built-in implementation currently, but is far from guaranteed. This should be fixed too.

This task is to design a test strategy that both avoids this issue, and ensure good testing, including of the fact that otherwise valid calls fail when the follow a call the return an error (something that's not systematically tested so far), and implement it.

Cc @gilles-peskine-arm who might expand on this description by suggesting reasonable strategies.

### PR DESCRIPTION
**Pull request by valeriosetti**
## Description

As explained in issue #6502, JPAKE tests (especially the ones using `ecjpake_setup()` test function) rely on the possibility to successfully re-use context's structures even after failing commands or "manually" duplicating them. Even though this works for the current built-in implementation, this is not guaranteed by the PSA specs, so it should be avoided.

Therefore the goal of this PR is to address this problem. This has been achieved by means of:

- changes to `psa_crypto_pake.c` functions
    - this is to enforce PAKE context's compromise even for the built-in implementation in case of errors
    - in this way all the "old" tests which were based on wrong assumptions fail
- changes to `test_suite_psa_crypto.function` and `test_suite_psa_crypto.data` files
    - in order to adapt tests to the new format

This PR resolves #6502.

## Gatekeeper checklist

- [ ] **changelog** not required - tests only
- [ ] **backport** not required - new feature
- [ ] **tests** provided

### CODE REVIEW

**Comment by mpg** [comment]
It looks like some changes where included by mistake in the last commit - most of `test_suite_psa_crypto.data` being commented out. I'll review once this is fixed.

By the way, I've been wondering if we should actually move the PAKE tests to their own files `test_suite_psa_crypto_pake.{data,function}`, which would avoid such issues. @gilles-peskine-arm wdyt?

**Comment by valeriosetti** [comment]
> It looks like some changes where included by mistake in the last commit - most of test_suite_psa_crypto.data being commented out. I'll review once this is fixed.

Ops, sorry, I did that to speed up development and focus only on relevant part for this task. I'll fix this immediately

> By the way, I've been wondering if we should actually move the PAKE tests to their own files test_suite_psa_crypto_pake.{data,function}, which would avoid such issues

FWIW it makes sense to me. It helps reading/debugging the `.data` file

**Comment by gilles-peskine-arm** [comment]
> By the way, I've been wondering if we should actually move the PAKE tests to their own files test_suite_psa_crypto_pake.{data,function}, which would avoid such issues. @gilles-peskine-arm wdyt?

[`test_suite_psa_crypto` is too big and should be split](https://github.com/Mbed-TLS/mbedtls/issues/4967). So yes, please feel free to create new test suites.

> most of test_suite_psa_crypto.data being commented ou

Note that if you want to run just one (or a small number) of tests for debugging, you can copy them to a new `.data` file with the same first part of the name, e.g. create a `test_suite_psa_crypto.tmp.data`, and running `make` will build the corresponding `test_suite_psa_crypto.tmp` with just those test cases.

**Comment by gilles-peskine-arm** [comment]
FYI CI is failing on the current head.

**Comment by mpg** [comment]
> Note that if you want to run just one (or a small number) of tests for debugging, you can copy them to a new `.data` file with the same first part of the name, e.g. create a `test_suite_psa_crypto.tmp.data`, and running `make` will build the corresponding `test_suite_psa_crypto.tmp` with just those test cases.

Thanks for the tip, I didn't know. I usually cut most of the `.data` file and then have to remember to paste it back before committing - so I wasn't very surprised when I saw the accidental commit here :)

**Comment by valeriosetti** [comment]
> @valeriosetti This is label needs-work, is that still correct or is it ready for review? I see the CI's green now.

Sorry, I forgot to adjust labels after fixing everything

File: tests/suites/test_suite_psa_crypto.function [hunk 1]
```diff
@@ -8779,54 +8791,90 @@ void ecjpake_setup( int alg_arg, int key_type_pw_arg, int key_usage_pw_arg,
...
+                    expected_error );
+        PSA_ASSERT( psa_pake_abort( &operation ) );
+        goto exit;
+    }
 
-    TEST_EQUAL( psa_pake_setup( &operation, &cipher_suite ),
-                expected_status_setup );
-    if( expected_status_setup != PSA_SUCCESS )
+    status = psa_pake_setup( &operation, &cipher_suite );
+    if (status != PSA_SUCCESS)
```
Comment (mpg) [inline tests/suites/test_suite_psa_crypto.function]: This doesn't look correct: basically we're asserting that `status` is either success or the expected error. But we should know whether we expect success or not at this stage and assert that. I mean currently, imagine we invoke the function with `INJECT_ERR_DUPLICATE_SETUP, PSA_ERROR_BAD_STATE` as the last two arguments. The `psa_pake_setup()` is expected to return success, but currently if it returns `PSA_ERROR_BAD_STATE` for some reason the test will still be marked as passed.

The underlying issue is that we have a single `expected_error` variable for all functions that may fail. We should either have multiple variables (`expected_status_setup`, `expected_status_set_role`, etc) or an extra variable `expecting_error_at_stage_xxx`, taking value from a new enum, or any other way to ensure that only the function that we expect to fail actually fails.

File: tests/suites/test_suite_psa_crypto.function [hunk 2]
```diff
@@ -8779,54 +8791,90 @@ void ecjpake_setup( int alg_arg, int key_type_pw_arg, int key_usage_pw_arg,
...
-                                 NULL, 0, NULL ),
-                PSA_ERROR_BAD_STATE );
-    TEST_EQUAL( psa_pake_input( &operation, PSA_PAKE_STEP_KEY_SHARE, NULL, 0),
-                PSA_ERROR_BAD_STATE );
-
-    PSA_ASSERT( psa_pake_abort( &operation ) );
+    if ( inj_err_type == INJECT_ERR_UNINITIALIZED_ACCESS )
+    {
+        TEST_EQUAL( psa_pake_set_user( &operation, NULL, 0 ),
+                    expected_error );
```
Comment (mpg) [inline tests/suites/test_suite_psa_crypto.function]: I'm not sure if we should use `expected_error` or just hardcode `PSA_ERR_BAD_STATE`. Since this is an "internal" error (as the commit message calls it), it would seem logical for the expected value to also be internal rather than taken from the arguments? (That's what's done for `INJECT_ERR_INVALID_USER` and later, but not for the once before that.)

Probably doesn't matter much in practice, as hopefully we'll pass the correct value in the `.data` file anyway, but it's probably better for things to be consistent.

File: tests/suites/test_suite_psa_crypto.function (hunk 1 above)
Comment (mpg) [reply tests/suites/test_suite_psa_crypto.function]: Actually perhaps we can extend the existing enum use for injecting "internal" errors to indicate the stage for "external" errors as well:
```c
typedef enum
{
    ERR_NONE = 0,
    ERR_INJECT_UNINITIALIZED_ACCESS,
    /* ... */
    ERR_IN_SETUP,
    ERR_IN_SET_PASSWORD_KEY,
    /* ... */
} ecjpake_error_stage_t;
```
then when we pass arguments that will cause setup to fail, we also pass `ERR_IN_SETUP`, etc.

And here we would do
```c
if( error_stage != ERR_IN_SETUP )
{
    PSA_ASSERT( status );
}
else
{
    TEST_EQUAL( status, expected_error );
    goto exit;
}
```

Wdyt?

**Review by mpg — CHANGES_REQUESTED** [review]
Thanks for adding a lot of test cases :) I think the new structure of the test code is much better, but I'm afraid it still suffers from a fundamental issue where we may accept failures at other stages than we should.

While reading it, it made me realise that we can, and probably should improve `ecjpake_do_round()` with the same principles: currently is accepts a variable that say at which round if any we should accept errors. When injecting errors in one round, it always corrupts the same part - the two proofs. It would be better to only inject one error at a time, and make it configurable which part if corrupted (proof, public, key share, and when there are two of them the first on the second one). However that goes beyond the original goal of the this task, so if you feel it's scope creep, please feel free to push back and we can leave this for another PR.

**Comment by valeriosetti** [comment]
> However that goes beyond the original goal of the this task, so if you feel it's scope creep, please feel free to push back and we can leave this for another PR.

In my opinion it does not make much sense to open more than 1 task/PR to solve the same problem. It is reasonable that requirements are slightly adjusted during the review process. I´ll try to address them ;)

File: tests/suites/test_suite_psa_crypto_pake.function [hunk 3]
```diff
@@ -596,22 +596,57 @@ exit:
...
+        PSA_ASSERT( status );                                               \
+    }                                                                       \
+    else                                                                    \
+    {                                                                       \
+        TEST_EQUAL( status, expected_error );                               \
+        goto exit;                                                          \
+    }
 
+/*
+ * This check is used for errors injected explicitly. The check is conditional
```
Comment (mpg) [inline tests/suites/test_suite_psa_crypto_pake.function:640]: Hmm, the distinction between the two macros was not entirely clear to me initially. I think the bit that would have helped me get it faster is this: the first macro is used for code that may either succeed or fail, and this is why it's always called - in case of success, future steps depend on this one. So I think the fact that both comments start with "This check is used for errors" is slightly misleading, as only in the second case is an error always expected.

**Review by mpg — COMMENTED** [review]
LGTM, now just expecting an extension to `ecjpake_do_round()` as discussed before.

File: tests/suites/test_suite_psa_crypto_pake.function [hunk 4]
```diff
@@ -0,0 +1,1022 @@
...
+
+    psa_pake_cs_set_algorithm( &cipher_suite, alg );
+    psa_pake_cs_set_primitive( &cipher_suite, primitive_arg );
+    psa_pake_cs_set_hash( &cipher_suite, hash_alg );
+
+    /* Get shared key */
+    PSA_ASSERT( psa_key_derivation_setup( &server_derive, derive_alg ) );
+    PSA_ASSERT( psa_key_derivation_setup( &client_derive, derive_alg ) );
+
+    if( PSA_ALG_IS_TLS12_PRF( derive_alg ) ||
```
Comment (mprse) [inline tests/suites/test_suite_psa_crypto_pake.function:921]: Here we are checking that `derive_alg` is:`TLS12_PRF` or`TLS12_PSK_TO_MS`, but in the test cases only `TLS12_PSK_TO_MS` is used.

**Review by mprse — CHANGES_REQUESTED** [review]
I like how pake tests are organized now. PR looks good left few suggestions for improvements.

File: tests/suites/test_suite_psa_crypto_pake.function [hunk 5]
```diff
@@ -0,0 +1,1022 @@
...
+/*
+ * This check is used for errors issued through wrong input parameters. The
+ * check is always performed because, in case of all valid parameters, the
+ * setup function can go on
+ */
+#define SETUP_ALWAYS_CHECK_STEP( test_function, this_check_err_stage )      \
+    status = test_function;                                                 \
+    if( err_stage != this_check_err_stage )                                 \
+    {                                                                       \
+        PSA_ASSERT( status );                                               \
```
[the remark this answers is not in the document]
Comment (mpg) [reply tests/suites/test_suite_psa_crypto_pake.function:631]: `PSA_ASSERT()` uses `TEST_EQUAL()` so it will print the same information. (Perhaps you're confusing with `TEST_ASSERT()` which indeed should be avoided when better options are available.)

File: tests/suites/test_suite_psa_crypto_pake.function [hunk 6]
```diff
@@ -0,0 +1,1022 @@
...
+        psa_set_key_type( &attributes, key_type_pw );
+        PSA_ASSERT( psa_import_key( &attributes, pw_data->x, pw_data->len,
+                                    &key ) );
+    }
+
+    psa_pake_cs_set_algorithm( &cipher_suite, alg );
+    psa_pake_cs_set_primitive( &cipher_suite, primitive );
+    psa_pake_cs_set_hash( &cipher_suite, hash_alg );
+
+    PSA_ASSERT( psa_pake_abort( &operation ) );
```
[the remark this answers is not in the document]
Comment (mpg) [reply tests/suites/test_suite_psa_crypto_pake.function:700]: I think the purpose of this line is to check that `abort()` works on a freshly-initialized operation. Perhaps as comment is warranted. Or make it optional if possible so we test both that abort works here, but also that future calls work without the abort here.

File: tests/suites/test_suite_psa_crypto_pake.function (hunk 5 above)
[the remark this answers is not in the document]
Comment (mprse) [reply tests/suites/test_suite_psa_crypto_pake.function:631]: Oh, yes this is `PSA`. Thanks for clarification. Resolving this one.

File: tests/suites/test_suite_psa_crypto_pake.function (hunk 6 above)
[the remark this answers is not in the document]
Comment (mprse) [reply tests/suites/test_suite_psa_crypto_pake.function:700]: If this is intentional to verify abort on freshly-initialized operation, then we need another test case for this and then test setup and the rest of api without abort before.

File: tests/suites/test_suite_psa_crypto_pake.data [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+PSA PAKE: duplicate a valid setup
+depends_on:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:PSA_WANT_ECC_SECP_R1_256:PSA_WANT_ALG_SHA_256
+ecjpake_setup:PSA_ALG_JPAKE:PSA_KEY_TYPE_PASSWORD:PSA_KEY_USAGE_DERIVE:PSA_PAKE_PRIMITIVE(PSA_PAKE_PRIMITIVE_TYPE_ECC, PSA_ECC_FAMILY_SECP_R1, 256):PSA_ALG_SHA_256:PSA_PAKE_ROLE_SERVER:0:"abcd":ERR_INJECT_DUPLICATE_SETUP:PSA_ERROR_BAD_STATE
+
+PSA PAKE: ecjpake setup invalid role NONE
+depends_on:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:PSA_WANT_ECC_SECP_R1_256:PSA_WANT_ALG_SHA_256
+ecjpake_setup:PSA_ALG_JPAKE:PSA_KEY_TYPE_PASSWORD:PSA_KEY_USAGE_DERIVE:PSA_PAKE_PRIMITIVE(PSA_PAKE_PRIMITIVE_TYPE_ECC, PSA_ECC_FAMILY_SECP_R1, 256):PSA_ALG_SHA_256:PSA_PAKE_ROLE_NONE:0:"abcd":ERR_IN_SET_ROLE:PSA_ERROR_NOT_SUPPORTED
+
+PSA PAKE: wrong key type password
```
[the remark this answers is not in the document]
Comment (valeriosetti) [reply tests/suites/test_suite_psa_crypto_pake.data:33]: TBH I think I copy/pasted this part from the original `data` file without paying too much attention, but you're right. I'll fix this

File: tests/suites/test_suite_psa_crypto_pake.function (hunk 4 above)
Comment (valeriosetti) [reply tests/suites/test_suite_psa_crypto_pake.function:921]: Ok. I'll put this one together with the change suggested by @mpg [here](https://github.com/Mbed-TLS/mbedtls/pull/6601#pullrequestreview-1185874396)

File: tests/suites/test_suite_psa_crypto_pake.function [hunk 8]
```diff
@@ -3,12 +3,88 @@
...
+    PAKE_ROUND_ONE,
+    PAKE_ROUND_TWO
+} pake_round_t;
+
+/* Inject an error on the specified buffer ONLY it this is the correct stage */
+#define DO_ROUND_CONDITIONAL_INJECT( this_stage, buf ) \
+    if ( this_stage == err_stage )                  \
+    {                                               \
+        *( buf + 7) ^= 1;                           \
+        *( buf + 8 ) ^= 1;                          \
```
Comment (mpg) [inline tests/suites/test_suite_psa_crypto_pake.function:60]: Minor: I'd rather flip a single bit. I'm not sure why we flipped two of them before. Also, please add a comment that offset 7 is arbitrary but chosen because it's "in the middle" of the thing we're corrupting.

File: tests/suites/test_suite_psa_crypto_pake.function [hunk 9]
```diff
@@ -3,12 +3,88 @@
...
+    }
+
+#define DO_ROUND_UPDATE_OFFSETS( main_buf_offset, step_offset, step_size ) \
+    {                                       \
+        step_offset = main_buf_offset;      \
+        main_buf_offset += step_size;        \
+    }
+
+#define DO_ROUND_CHECK_FAILURE(  )                                  \
```
Comment (mpg) [inline tests/suites/test_suite_psa_crypto_pake.function:69]: Ah, thanks for making that a macro.

**Review by mpg — COMMENTED** [review]
Looking pretty good. Just two minor things.

**Comment by mpg** [comment]
The CI points out that some test case descriptions are not unique:
```
tests/suites/test_suite_psa_crypto_pake.data:192:ERROR:Duplicate description (also line 180)
tests/suites/test_suite_psa_crypto_pake.data:196:ERROR:Duplicate description (also line 184)
tests/suites/test_suite_psa_crypto_pake.data:200:ERROR:Duplicate description (also line 188)
```

File: tests/suites/test_suite_psa_crypto_pake.function (hunk 8 above)
Comment (valeriosetti) [reply tests/suites/test_suite_psa_crypto_pake.function:60]: If I remember correctly there were 2 bits flipped, but in two different parts. Therefore, now that we have the control at which step the error is introduced, it's more correct to reduce it to a single bit. I agree ;)

File: tests/suites/test_suite_psa_crypto_pake.data [hunk 10]
```diff
@@ -65,17 +57,10 @@ ecjpake_setup:PSA_ALG_JPAKE:PSA_KEY_TYPE_PASSWORD:PSA_KEY_USAGE_DERIVE:PSA_PAKE_
 # NOTE: this test should be enabled once the psa_pake_set_password_key() function
 #       will reject empty passwords. The expected error code must be adjusted
 #       accordingly to the code
-#PSA PAKE: empty server password
+#PSA PAKE: empty password
```
[the remark this answers is not in the document]
Comment (valeriosetti) [reply tests/suites/test_suite_psa_crypto_pake.data:60]: Ok, thanks for the explanation!
But if we don't need to test the empty password case (i.e. we don't test `psa_import_key()`, then I can remove the `"abcd"` password from the `data` file as input argument to `ecjpake_setup()`. Isn't it?

[the remark this answers is not in the document]
Comment (valeriosetti) [reply tests/suites/test_suite_psa_crypto_pake.data:60]: Or, otherwise, I can add a specific "injected error" for this case

[the remark this answers is not in the document]
Comment (mpg) [reply tests/suites/test_suite_psa_crypto_pake.data:60]: > But if we don't need to test the empty password case (i.e. we don't test `psa_import_key()`, then I can remove the `"abcd"` password from the `data` file as input argument to `ecjpake_setup()`. Isn't it?

Indeed, good point, we can simplify the function by removing this parameter.

(We don't need to test `psa_import_key()` here, it should have its own unit tests already.)

**Comment by mpg** [comment]
Relaying another complaint from the internal CI:
```
 88/104 Test  #88: psa_crypto_pake-suite ......................***Failed    0.01 sec
PSA PAKE: uninitialized access to psa_pake_operation_t ............ ==4722==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x4d8b39  (/var/lib/build/tests/test_suite_psa_crypto_pake+0x4d8b39)
    #1 0x4ee50f  (/var/lib/build/tests/test_suite_psa_crypto_pake+0x4ee50f)
    #2 0x487a0b  (/var/lib/build/tests/test_suite_psa_crypto_pake+0x487a0b)
    #3 0x488f52  (/var/lib/build/tests/test_suite_psa_crypto_pake+0x488f52)
    #4 0x4932cf  (/var/lib/build/tests/test_suite_psa_crypto_pake+0x4932cf)
    #5 0x494898  (/var/lib/build/tests/test_suite_psa_crypto_pake+0x494898)
    #6 0x7f392478983f  (/lib/x86_64-linux-gnu/libc.so.6+0x2083f)
    #7 0x419a08  (/var/lib/build/tests/test_suite_psa_crypto_pake+0x419a08)

SUMMARY: MemorySanitizer: use-of-uninitialized-value (/var/lib/build/tests/test_suite_psa_crypto_pake+0x4d8b39) 
Exiting
```
This is from `./tests/scripts/all.sh --seed 4 --keep-going test_memsan` but other memsan tests are failing (and valgrind as well, as expected).

For local investigation I would recommend using `cmake -D CMAKE_BUILD_TYPE=MemSanDbg` to build before running the test suite; that should turn the addresses above into something readable. Alternatively, build with make and set `CFLAGS` manually (can copy them `CMAKE_C_FLAGS_ASANDBG` in `CMakeLists.txt`), or do a debug build however you want on run the test suite under valgrind.

**Review by mpg — COMMENTED** [review]
Just a note that I've reviewed the last three commits.

**Comment by valeriosetti** [comment]
The branch was forced-pushed due to conflicts with `development`.
Before this HEAD was at `29d51128f06195c7b8106552774f095a54f0bf2b`

**Note**
Conflicting changes are not yet included in this branch because [some of them] must be reworked.
From a first quick analysis I would say that:

* to be adapted
    * cd356c3cdb312e276473e038d1593f6f92bcd5b3
* to be removed (since it's already addressed by this branch differently)
    * c018204019a6a935ae0bf3ab888cd7a6d7fc2039
    * 7c7954842b6e287f95168a5dafdc00f0491e1675
    * f82effa9826a0e93aaa8c4c7928ad1016a16a8e8

**Review by mpg — DISMISSED** [review]
LGTM

File: tests/suites/test_suite_psa_crypto_pake.function (hunk 6 above)
[the remark this answers is not in the document]
Comment (mprse) [reply tests/suites/test_suite_psa_crypto_pake.function:588]: This hasn't been changed but discussed situation is tested in `ecjpake_rounds_inject`.

**Review by mprse — CHANGES_REQUESTED** [review]
Thanks for addressing my comments.
Only one dependency in test case is invalid. Otherwise LGTM.

**Review by mprse — DISMISSED** [review]

File: tests/suites/test_suite_psa_crypto_pake.data [hunk 11]
```diff
@@ -99,7 +99,7 @@ depends_on:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:PSA_WANT_ECC_SECP_R1_256:PSA_WANT_ALG_
 ecjpake_rounds:PSA_ALG_JPAKE:PSA_PAKE_PRIMITIVE(PSA_PAKE_PRIMITIVE_TYPE_ECC, PSA_ECC_FAMILY_SECP_R1, 256):PSA_ALG_SHA_256:PSA_ALG_TLS12_PSK_TO_MS(PSA_ALG_SHA_256):"abcdef":0:0:ERR_NONE
 
 PSA PAKE: check rounds w/o forced errors, TLS12_PRF
-depends_on:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:PSA_WANT_ECC_SECP_R1_256:PSA_WANT_ALG_SHA_256:PSA_WANT_ALG_TLS12_PSK_TO_MS
+depends_on:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:PSA_WANT_ECC_SECP_R1_256:PSA_WANT_ALG_SHA_256:A_WANT_ALG_TLS12_PRF
```
Comment (mpg) [inline tests/suites/test_suite_psa_crypto_pake.data:102]: Missing 2 letters? `A_WANT_` -> `PSA_WANT_`

**Review by mpg — CHANGES_REQUESTED** [review]
Unfortunate typo in the last commit. (Feel free to amend the last commit and force-push.)

File: tests/suites/test_suite_psa_crypto_pake.data (hunk 11 above)
Comment (valeriosetti) [reply tests/suites/test_suite_psa_crypto_pake.data:102]: Ops, that's embarrassing, sorry...
I thought: "well it's only a single line to be changed, there's no need to test; what could possibly go wrong?" :/

**Review by mprse — APPROVED** [review]

File: tests/suites/test_suite_psa_crypto_pake.data (hunk 11 above)
Comment (mpg) [reply tests/suites/test_suite_psa_crypto_pake.data]: Yeah, I know the feeling :)

Actually just running the test suite wouldn't have caught it without taking a detailed look at the result: the test would have been skipped. I don't think it would have been caught by the CI either. We do have plans for this to be caught in the future, though. https://github.com/Mbed-TLS/mbedtls/issues/2691

**Review by mpg — APPROVED** [review]
LGTM
